lock-impl/src/test/java/com/palantir/lock/impl/ThreadInfoLockServiceImplTest.java

```diff
@@ -0,0 +1,308 @@
+            .isStandaloneServer(false)
+            .threadInfoConfiguration(ImmutableDebugThreadInfoConfiguration.builder()
+                    .recordThreadInfo(false)
+                    .build())
+            .build());
+
+    @Test
+    public void initialThreadInfoIsEmpty() {
+        forceSnapshot();
+        assertThat(getLatestThreadInfoForLock(TEST_LOCK_1)).isNull();
+        assertThat(getLatestThreadInfoForLock(TEST_LOCK_2)).isNull();
+        assertThat(getLatestThreadInfoForLock(TEST_LOCK_3)).isNull();
+    }
+
+    @Test
+    public void recordsThreadInfoForLock() throws InterruptedException {
+        lockService.lockWithFullLockResponse(LockClient.ANONYMOUS, LOCK_1_THREAD_1_WRITE_REQUEST);
+        forceSnapshot();
+
+        assertThat(getLatestThreadInfoForLock(TEST_LOCK_1)).isEqualTo(ANONYMOUS_TEST_THREAD_1);
+    }
+
+    @Test
+    public void recordsThreadInfoForUnlock() throws InterruptedException {
+        LockResponse response =
+                lockService.lockWithFullLockResponse(LockClient.ANONYMOUS, LOCK_1_THREAD_1_WRITE_REQUEST);
+        forceSnapshot();
+
+        assertThat(getLatestThreadInfoForLock(TEST_LOCK_1)).isEqualTo(ANONYMOUS_TEST_THREAD_1);
+
+        lockService.unlock(response.getToken());
+        forceSnapshot();
+
+        assertThat(getLatestThreadInfoForLock(TEST_LOCK_1)).isNull();
+    }
+
+    @Test
+    public void recordsThreadInfoForMultipleLocks() throws InterruptedException {
+        int numThreads = 10;
+        int numLocksPerThread = 10;
+
+        List<String> threadNames =
+                IntStream.range(0, numThreads).mapToObj(i -> "test-thread-" + i).collect(Collectors.toList());
+        Map<String, Set<LockDescriptor>> locksPerThread = new HashMap<>();
+        for (String threadName : threadNames) {
+            Set<LockDescriptor> locks = IntStream.range(0, numLocksPerThread)
+                    .mapToObj(i -> StringLockDescriptor.of("test-lock-" + i + "-from-thread-" + threadName))
+                    .collect(Collectors.toSet());
+            locksPerThread.put(threadName, locks);
+        }
+
+        for (String threadName : threadNames) {
+            LockRequest lockRequest = LockRequest.builder(
+                            ImmutableSortedMap.copyOf(locksPerThread.get(threadName).stream()
+                                    .collect(Collectors.toMap(lock -> lock, lock -> LockMode.WRITE))))
+                    .withCreatingThreadName(threadName)
+                    .build();
+            lockService.lockWithFullLockResponse(LockClient.ANONYMOUS, lockRequest);
+        }
+
+        forceSnapshot();
+
+        locksPerThread.forEach((threadName, lockDescriptors) ->
+                lockDescriptors.forEach(lock -> assertThat(getLatestThreadInfoForLock(lock))
+                        .isEqualTo(LockClientAndThread.of(LockClient.ANONYMOUS, threadName))));
+    }
+
+    @Test
+    public void recordsThreadInfoForSharedLock() throws InterruptedException {
```

Comment:
A lot of the tests from here and below are arguably overkill, given the semantics of the snapshotter - this is more testing the semantics of who acquires locks, as opposed to anything relevant to the snapshotter.

But for now, just leave them here.

Reply:
I agree that with the current snapshotter implementation, they are mostly testing the semantics of the lock service. They were crucial for testing alternative approaches (e.g. maintaining the info on every request) so I would suggest keeping them around in case we make changes to this feature.